packages/desktop/src/protocol.rs

```diff
@@ -51,12 +63,132 @@ fn module_loader(root_name: &str) -> String {
     )
 }
 
-pub(super) fn desktop_handler(
+/// An arbitrary asset is an HTTP response containing a binary body.
+pub type AssetResponse = Response<Cow<'static, [u8]>>;
+
+/// A future that returns an [`AssetResponse`]. This future may be spawned in a new thread,
+/// so it must be [`Send`], [`Sync`], and `'static`.
+pub trait AssetFuture: Future<Output = Option<AssetResponse>> + Send + Sync + 'static {}
+impl<T: Future<Output = Option<AssetResponse>> + Send + Sync + 'static> AssetFuture for T {}
+
+/// A handler that takes an asset [`Path`] and returns a future that loads the path.
+/// This handler is stashed indefinitely in a context object, so it must be `'static`.
+pub trait AssetHandler<F: AssetFuture>: Fn(&Path) -> F + Send + Sync + 'static {}
+impl<F: AssetFuture, T: Fn(&Path) -> F + Send + Sync + 'static> AssetHandler<F> for T {}
+
+/// An identifier for a registered asset handler, returned by [`AssetHandlerRegistry::register_handler`].
+#[derive(PartialEq, Eq, Hash, Clone, Copy, Debug)]
+pub struct AssetHandlerId(usize);
+
+struct AssetHandlerRegistryInner {
+    handlers: HashMap<
```

Comment:
This looks a lot like a [slab](https://docs.rs/slab/latest/slab/). If we don't need the AssetHandlerId to be unique throughout time, it might be simpler to use a slab directly